pkg/api/api.go

```diff
@@ -68,6 +76,86 @@ func (a *ScratchDataAPIStruct) AuthGetDatabaseID(ctx context.Context) int64 {
 	return ctx.Value("databaseId").(int64)
 }
 
+func (a *ScratchDataAPIStruct) CreateQuery(w http.ResponseWriter, r *http.Request) {
+	apiKey := r.URL.Query().Get("api_key")
+	if apiKey != "local" { // Validate API key
+		w.WriteHeader(http.StatusUnauthorized)
+		w.Write([]byte("Unauthorized"))
+		return
+	}
+
+	var requestBody struct {
+		Query    string `json:"query"`
+		Duration int    `json:"duration"` // Duration in seconds
+	}
+
+	if err := json.NewDecoder(r.Body).Decode(&requestBody); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte("Invalid request body"))
+		return
+	}
+
+	// Validate the query
+	if requestBody.Query == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte("Query cannot be empty"))
+		return
+	}
+
+	// Generate a new UUID
+	queryUUID := uuid.New()
+
+	// Store the query and its expiration time
+	queryExpiration := time.Duration(requestBody.Duration)
+	a.cache.Set(queryUUID.String(), []byte(requestBody.Query), &queryExpiration)
+
+	// Return the UUID representing the query
+	w.Header().Set("Content-Type", "application/json")
+	response := struct {
+		QueryUUID string `json:"query_uuid"`
+	}{
+		QueryUUID: queryUUID.String(),
+	}
+	json.NewEncoder(w).Encode(response)
+}
+
+func (a *ScratchDataAPIStruct) executeQueryAndStreamData(w http.ResponseWriter, ctx context.Context, query string, databaseID int64, format string) error {
+	dest, err := a.destinationManager.Destination(databaseID)
+	if err != nil {
+		return err
+	}
+
+	switch strings.ToLower(format) {
+	case "csv":
+		w.Header().Set("Content-Type", "text/csv")
+		return dest.QueryCSV(query, w)
+	default:
+		w.Header().Set("Content-Type", "application/json")
+		return dest.QueryJSON(query, w)
+	}
+}
+
+func (a *ScratchDataAPIStruct) ShareData(w http.ResponseWriter, r *http.Request) {
+	queryUUID := chi.URLParam(r, "uuid")
+	format := chi.URLParam(r, "format")
+
+	// Retrieve query from cache using UUID
+	query, found := a.cache.Get(queryUUID)
+	if !found {
+		http.Error(w, "Query not found", http.StatusNotFound)
+		return
+	}
+
+	// Convert query to string
+	queryStr := string(query)
+
+	// Execute query and stream data
+	databaseID := a.AuthGetDatabaseID(r.Context())
```

Comment:
`databaseID` should be retrieved from the cache. It should not come from `r.Context()`. That is because the query is specific to a database. With the current code, a saved query could be run against a different database, which would be an error.